dcpy/connectors/edm/publishing.py

```diff
@@ -0,0 +1,42 @@
+from datetime import date
+from pathlib import Path
+
+from dcpy.connectors import s3
+from dcpy.git import git_branch
+
+BUCKET = "edm-publishing"
+
+
+def upload(
+    output: Path,
+    publishing_folder: str,
+    acl: str,
+    *,
+    s3_subpath: str = None,
```

Comment:
Do we have an established enough folder convention on edm-publishing to just infer the `s3_subpath` from, say, the "name" of the release, then potentially a version of it? 

And for an example of what I mean by "name," in `edm-publishing/ db-pluto / 2020-02-23 / output` it would be  "2020-02-03". Then some releases every now and then have a subfolder (e.g. `edm-publishing/ db-cbbr / fvankrieken-patch-pluto-gha / latest / output`)

I think that, ideally, a user of this module would never have to think about S3 paths

Reply:
Yeah that's a good point, this really does need "version" (or "name" as you're referring to it) should be an argument, defaulting to current date is certainly not valid for all datasets. And I guess to your point, a "branch name" or something along those lines COULD just be a version, though often within dev branches it can still be useful to generate multiple versions, but even then current defaults don't necessarily fully meet needs (if "version" is just date, can't generate two outputs in the same day). 

Tying in to your below comment here as well - our most consistent norm with output folders now is `{dataset}/{branch}/{version}/...` so wanted to go with that as a sort of default.

PFF has some oddness in that there are sort of "subproducts", but I guess a simple way to handle that would be just have "publishing_folder" be `db-factfinder/acs` etc. But I do sort of like the norm of just a few arguments that are just single folder names rather than paths and follow some sort of default structure, and then some optional arg the one I have which just gives complete control over folder path

Reply:
Or actually for PFF, this is my simplified version of the folder structure - it used to have another that seems redundant

```python
    upload(
        output=file,
        publishing_folder=f"db-factfinder",
        version=str(date.today()),
        acl="public-read",
        s3_subpath=Path(git_branch()) / export_type / year,
        latest=latest,
    )
```

Reply:
So in my mind, we sort of have the more "absolute" norms of top-level publishing folder and bottom level version folder, but in between can be somewhat custom. Definitely open to simplifying it in some way. @damonmcc your thoughts on all this?

Reply:
I don't think we have or should impose a folder convention right now, but the most common pattern does seem to be `{dataset}/{branch}/{version}/...`. 

per our discussion issue on output folder naming (https://github.com/NYCPlanning/data-engineering/issues/98) and chats we've had about "should we have to build on main for a production release?", I'm not sure we'll like having `branch` be such a high-level part of the folder convention forever

TLDR: I'm in favor of flexibility for now and deciding on a convention later

Reply:
@alexrichey I think we'll go that route for now, leaving an optional subpath with no default just to make sure that fine-grained control is allowed while still keeping the more reliable/consistent "publishing folder" and "version" args